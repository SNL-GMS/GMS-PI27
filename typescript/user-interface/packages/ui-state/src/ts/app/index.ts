export * from './api';
export * from './history';
export * from './hooks';
export * from './loader';
export * from './processing';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
