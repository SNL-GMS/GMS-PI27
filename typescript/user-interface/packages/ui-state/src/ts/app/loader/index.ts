export * from './actions';
export * from './loader-reducer';
