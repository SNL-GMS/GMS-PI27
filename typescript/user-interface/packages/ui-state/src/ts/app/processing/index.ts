export * from './beamforming';
export * from './input-channels-by-prioritization';
export * from './validate-channels';
