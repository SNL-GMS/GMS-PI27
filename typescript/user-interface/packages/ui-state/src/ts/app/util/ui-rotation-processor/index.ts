export * from './rotate-2d-for-channels';
export * from './rotate-2d-for-signal-detections';
export * from './rotate-2d-for-stations';
export * from './ui-rotation-processor-utils';
