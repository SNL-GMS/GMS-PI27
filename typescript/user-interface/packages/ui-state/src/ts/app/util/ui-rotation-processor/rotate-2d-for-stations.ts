import type { CommonTypes } from '@gms/common-model';
import type { SamplingType } from '@gms/common-model/lib/common/types';
import type { Channel } from '@gms/common-model/lib/station-definitions/channel-definitions/channel-definitions';
import type { Station } from '@gms/common-model/lib/station-definitions/station-definitions/station-definitions';
import { UILogger } from '@gms/ui-util';
import flatMap from 'lodash/flatMap';
import React from 'react';

import type { MaskAndRotate2dResult } from '../../../types';
import { addRotatedChannelsAndChannelSegments } from '../../api/data/channel/channel-reducer';
import {
  useAppDispatch,
  useAppSelector,
  useChannels,
  useProcessingAnalystConfiguration,
  useStageId
} from '../../hooks';
import { selectOpenEvent, selectOpenIntervalName, selectUsername } from '../../state';
import { useGet2dRotationForSingleChannelPair } from './rotate-2d-for-channels';
import type { SDHypothesisArgs } from './ui-rotation-processor-utils';
import {
  extractRotationResultsFromPromises,
  getChannelPairsToRotate,
  useGetOrFetchRotationTemplateForStationAndPhase
} from './ui-rotation-processor-utils';

const logger = UILogger.create(
  'GMS_LOG_UI_ROTATION_PROCESSOR_STATION',
  process.env.GMS_LOG_UI_ROTATION_PROCESSOR_STATION
);

/**
 * higher order function that creates a function for rotate 2d stations
 *
 * @returns a function to perform 2D rotation for the given stations
 */
export function useRotate2dForStations() {
  const getOrFetchRotationTemplate = useGetOrFetchRotationTemplateForStationAndPhase();
  const get2dRotationForSingleChannelPair = useGet2dRotationForSingleChannelPair();
  const allChannels = useChannels();
  const dispatch = useAppDispatch();
  const rotationReplacementAzimuthToleranceDeg =
    useProcessingAnalystConfiguration()?.rotation?.rotationReplacementAzimuthToleranceDeg;
  const username = useAppSelector(selectUsername);
  const openIntervalName = useAppSelector(selectOpenIntervalName);
  const stageId = useStageId();
  const currentOpenEvent = useAppSelector(selectOpenEvent);

  return React.useCallback(
    /**
     * Used to rotate channel pairs for all provided stations based on the phase rotation definition
     *
     * @param stations list of stations to rotate
     * @param phaseType selected phase type
     * @param samplingType optional sampling type
     * @param leadDuration optional lead duration
     * @param duration optional duration
     * @param location optional location
     * @param receiverToSourceAzimuthDeg optional receiver to source azimuth
     */
    async function rotate2dForStations(
      stations: Station[],
      phaseType: string,
      samplingType?: SamplingType,
      leadDuration?: number,
      duration?: number,
      location?: CommonTypes.Location,
      receiverToSourceAzimuthDeg?: number
    ) {
      if (!stageId) {
        throw new Error('No stage ID found, open interval before rotating');
      }
      const flattenedRotatedChannelPromises: Promise<MaskAndRotate2dResult[]>[] = stations.flatMap(
        async station => {
          // The InteractiveAnalysisStateManager collects the information it needs to perform 2D waveform rotation:
          // The InteractiveAnalysisStateManager selects the RotationTemplate for the Station and PhaseType literal.
          const rotationTemplate = await getOrFetchRotationTemplate(station, phaseType);

          // The InteractiveAnalysisStateManager selects from the UiStateStore the Channel pairs to rotate.
          const channelsToRotate: [Channel, Channel][] = getChannelPairsToRotate(
            allChannels,
            station,
            rotationTemplate
          );

          // If the leadDuration and duration values are provided to the operation then use these values for leadDuration and duration.
          // Otherwise use the leadDuration and duration values from the RotationTemplate.
          const leadDurationForStations = leadDuration ?? rotationTemplate.leadDuration;
          const durationForStations = duration ?? rotationTemplate.duration;

          // The InteractiveAnalysisStateManager calls the operation rotate2dForChannels(...) for each of the Channel pairs to rotate,
          // providing the Channels, the phaseType, the samplingType, the leadDuration, the duration, the location, and the
          // receiverToSourceAzimuthDeg.
          const rotatedChannelPromises: MaskAndRotate2dResult[][] = await Promise.all(
            channelsToRotate.flatMap(async channelPair =>
              get2dRotationForSingleChannelPair(
                channelPair,
                phaseType,
                samplingType,
                leadDurationForStations,
                durationForStations,
                location,
                receiverToSourceAzimuthDeg
              )
            )
          );
          return flatMap(rotatedChannelPromises);
        }
      );

      const rotationResultsArray: MaskAndRotate2dResult[] = await Promise.allSettled(
        flattenedRotatedChannelPromises
      )
        .then(promises => extractRotationResultsFromPromises(promises))
        .catch(error => {
          logger.error(error);
          return [];
        });

      const newSDHypothesisArgs: SDHypothesisArgs = {
        username,
        openIntervalName,
        stageId,
        currentEventId: currentOpenEvent?.id
      };
      dispatch(
        addRotatedChannelsAndChannelSegments({
          rotationResults: rotationResultsArray,
          config: {
            rotationReplacementAzimuthToleranceDeg
          },
          newSDHypothesisArgs
        })
      );
    },
    [
      username,
      openIntervalName,
      stageId,
      currentOpenEvent?.id,
      dispatch,
      rotationReplacementAzimuthToleranceDeg,
      getOrFetchRotationTemplate,
      allChannels,
      get2dRotationForSingleChannelPair
    ]
  );
}
