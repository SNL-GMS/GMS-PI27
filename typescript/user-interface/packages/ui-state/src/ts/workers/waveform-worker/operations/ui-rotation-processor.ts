import type { WaveformTypes } from '@gms/common-model';
import { UNFILTERED } from '@gms/common-model/lib/filter';
import { ChannelOrientationType } from '@gms/common-model/lib/station-definitions/channel-definitions/channel-definitions';
import { UILogger } from '@gms/ui-util';
import { maskAndRotate2d as wasmMaskAndRotate2d } from '@gms/ui-wasm';

import { createRotated } from '../../../app/util/channel-factory';
import { getChannelNameComponents } from '../../../app/util/channel-factory-util';
import {
  adjustRotationDefinitionForChannelOrientation,
  createRotatedUiChannelSegment,
  findOverlappingProcessingMasks
} from '../../../app/util/util';
import type { MaskAndRotate2dParams, MaskAndRotate2dResult } from '../../../types';
import { calculateAndStorePositionBuffer, generateUniqueId } from '../util/position-buffer-util';
import { hydrateTrimmedChannelSegmentSamples } from '../util/sample-util';

const logger = UILogger.create('GMS_UI_ROTATION_PROCESSOR', process.env.GMS_UI_ROTATION_PROCESSOR);

export const maskAndRotate2d = async ({
  rotationDefinition,
  station,
  channels,
  uiChannelSegmentPair,
  rotationTimeInterval,
  processingMasks,
  maskTaperDefinition
}: MaskAndRotate2dParams): Promise<MaskAndRotate2dResult[]> => {
  const hydratedChannelSegments = await Promise.all(
    uiChannelSegmentPair.map(async cs =>
      hydrateTrimmedChannelSegmentSamples(cs.channelSegment, cs.domainTimeRange)
    )
  );

  const missingDataFound = hydratedChannelSegments.some(cs => {
    return (
      cs.timeseries.reduce((result, ts) => {
        return result + (ts?.samples?.length || 0);
      }, 0) <= 0
    );
  });

  if (missingDataFound) {
    const { stationName } = getChannelNameComponents(
      uiChannelSegmentPair[0].channelSegment.id.channel.name
    );
    // Should be caught in toast
    throw new Error(`Failed to rotate station ${stationName}, is missing waveform data.`);
  }

  const waveforms = await wasmMaskAndRotate2d(
    rotationDefinition,
    [hydratedChannelSegments[0], hydratedChannelSegments[1]],
    rotationTimeInterval.startTimeSecs,
    rotationTimeInterval.endTimeSecs,
    processingMasks,
    maskTaperDefinition
  );

  const radialTimeseries = waveforms[0].timeseries;
  const radialMissingInputChannels = waveforms[0].missingInputChannels;
  const transverseTimeseries = waveforms[1].timeseries;
  const transverseMissingInputChannels = waveforms[1].missingInputChannels;

  const radialRotationDefinition = adjustRotationDefinitionForChannelOrientation(
    rotationDefinition,
    ChannelOrientationType.RADIAL
  );
  const radialChannel = await createRotated(
    channels,
    radialRotationDefinition,
    ChannelOrientationType.RADIAL
  );

  const transverseRotationDefinition = adjustRotationDefinitionForChannelOrientation(
    rotationDefinition,
    ChannelOrientationType.TRANSVERSE
  );
  const transverseChannel = await createRotated(
    channels,
    transverseRotationDefinition,
    ChannelOrientationType.TRANSVERSE
  );

  const radialProcessingMasks = findOverlappingProcessingMasks(
    uiChannelSegmentPair[0].channelSegment.maskedBy,
    radialTimeseries
  );

  const transverseProcessingMasks = findOverlappingProcessingMasks(
    uiChannelSegmentPair[1].channelSegment.maskedBy,
    transverseTimeseries
  );

  const results: MaskAndRotate2dResult[] = [
    createRotatedUiChannelSegment(
      radialChannel,
      rotationTimeInterval,
      uiChannelSegmentPair[0].domainTimeRange,
      radialTimeseries,
      uiChannelSegmentPair[0].channelSegment.timeseriesType,
      radialProcessingMasks,
      radialMissingInputChannels
    ),
    createRotatedUiChannelSegment(
      transverseChannel,
      rotationTimeInterval,
      uiChannelSegmentPair[0].domainTimeRange,
      transverseTimeseries,
      uiChannelSegmentPair[1].channelSegment.timeseriesType,
      transverseProcessingMasks,
      transverseMissingInputChannels
    )
  ].map(uiChannelSegment => {
    // Strip samples from the timeseries and store the waveforms
    const timeseries: WaveformTypes.Waveform[] = uiChannelSegment.channelSegment.timeseries.map(
      wave => {
        const id = generateUniqueId(uiChannelSegment.channelSegment, wave, rotationTimeInterval);
        // We don't need to wait for the result
        calculateAndStorePositionBuffer(id, wave, uiChannelSegment.domainTimeRange).catch(
          logger.error
        );

        return {
          startTime: wave.startTime,
          endTime: wave.endTime,
          type: wave.type,
          sampleCount: wave.sampleCount,
          sampleRateHz: wave.sampleRateHz,
          _uiClaimCheckId: id
        };
      }
    );

    return {
      stationName: station.name,
      phase: rotationDefinition.rotationDescription.phaseType,
      rotatedChannel:
        uiChannelSegment.channelSegmentDescriptor.channel.name === radialChannel.name
          ? radialChannel
          : transverseChannel,
      rotatedUiChannelSegment: {
        ...uiChannelSegment,
        channelSegment: {
          ...uiChannelSegment.channelSegment,
          timeseries,
          _uiFilterId: UNFILTERED
        }
      }
    };
  });

  return results;
};
