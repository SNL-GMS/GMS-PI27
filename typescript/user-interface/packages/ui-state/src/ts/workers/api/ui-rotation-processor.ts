import type { ChannelTypes, ProcessingMaskDefinitionTypes, WaveformTypes } from '@gms/common-model';
import type { ProcessingMask } from '@gms/common-model/lib/channel-segment';
import type { TimeRange } from '@gms/common-model/lib/common';
import type { RotationDefinition } from '@gms/common-model/lib/rotation/types';
import type { Station } from '@gms/common-model/lib/station-definitions/station-definitions/station-definitions';

import type { MaskAndRotate2dResult, UiChannelSegmentsPair } from '../../types';
import { WorkerOperations } from '../waveform-worker/operations';
import { waveformWorkerRpc } from '../worker-rpcs';

export const maskAndRotate2d = async (
  rotationDefinition: RotationDefinition,
  station: Station,
  channels: ChannelTypes.Channel[],
  uiChannelSegmentPair: UiChannelSegmentsPair<WaveformTypes.Waveform>,
  rotationTimeInterval: TimeRange,
  processingMasks: Record<string, ProcessingMask[]>,
  maskTaperDefinition?: ProcessingMaskDefinitionTypes.TaperDefinition
): Promise<MaskAndRotate2dResult[]> =>
  waveformWorkerRpc.rpc(WorkerOperations.MASK_AND_ROTATE_2D, {
    rotationDefinition,
    station,
    channels,
    uiChannelSegmentPair,
    rotationTimeInterval,
    processingMasks,
    maskTaperDefinition
  });
