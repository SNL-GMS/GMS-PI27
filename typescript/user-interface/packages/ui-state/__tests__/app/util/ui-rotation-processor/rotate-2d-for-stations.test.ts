import { type CommonTypes, WorkflowTypes } from '@gms/common-model';
import {
  akasgBHEChannel,
  akasgBHNChannel,
  akasgVersionReference,
  defaultStations,
  eventData2,
  openIntervalName,
  processingAnalystConfigurationData,
  rotationTemplate,
  rotationTemplateByPhaseByStationRecord
} from '@gms/common-model/__tests__/__data__';
import type { Channel } from '@gms/common-model/lib/station-definitions/channel-definitions/channel-definitions';
import { renderHook } from '@testing-library/react-hooks';
import cloneDeep from 'lodash/cloneDeep';
import type { AnyAction } from 'redux';
import type { MockStoreCreator } from 'redux-mock-store';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import type { AppState } from '../../../../src/ts/app';
import { AsyncActionStatus, userSessionActions, workflowActions } from '../../../../src/ts/app';
import { useRotate2dForStations } from '../../../../src/ts/app/util/ui-rotation-processor/rotate-2d-for-stations';
import { useGetOrFetchRotationTemplateForStationAndPhase } from '../../../../src/ts/app/util/ui-rotation-processor/ui-rotation-processor-utils';
import { useQueryStateResult } from '../../../__data__';
import { appState, getTestReduxWrapper } from '../../../test-util';
import { activityInterval } from '../../api/workflow/sample-data';

// eslint-disable-next-line no-console
console.error = jest.fn();
// eslint-disable-next-line no-console
console.warn = jest.fn();

jest.mock('../../../../src/ts/app/util/ui-rotation-processor/rotate-2d-for-channels', () => {
  const actual = jest.requireActual(
    '../../../../src/ts/app/util/ui-rotation-processor/rotate-2d-for-channels'
  );
  // need to return undefined until function is implemented
  const mockRotate2dForChannels = jest.fn(() => {
    return undefined;
  });
  return {
    ...actual,
    useRotate2dForChannels: jest.fn(() => mockRotate2dForChannels),
    useGet2dRotationForSingleChannelPair: jest.fn(() => mockRotate2dForChannels)
  };
});
jest.mock('../../../../src/ts/app/util/ui-rotation-processor/ui-rotation-processor-utils', () => {
  const actual = jest.requireActual(
    '../../../../src/ts/app/util/ui-rotation-processor/ui-rotation-processor-utils'
  );
  const mockGetOrFetchRotationTemplateForStationAndPhase = jest.fn(() => rotationTemplate);
  const channelsToRotate: [Channel, Channel][] = [[akasgBHEChannel, akasgBHNChannel]];
  // need to return undefined until function is implemented
  return {
    ...actual,
    getChannelPairsToRotate: jest.fn(() => channelsToRotate),
    useGetOrFetchRotationTemplateForStationAndPhase: jest.fn(
      () => mockGetOrFetchRotationTemplateForStationAndPhase
    )
  };
});

jest.mock('axios', () => {
  const success = 'success';
  const actualAxios = jest.requireActual('axios');
  return {
    ...actualAxios,
    request: jest.fn().mockReturnValue(Promise.resolve(success))
  };
});

jest.mock('../../../../src/ts/app/api/data/signal-enhancement', () => {
  const actual = jest.requireActual('../../../../src/ts/app/api/data/signal-enhancement');
  return {
    ...actual,
    getRotationTemplates: jest.fn()
  };
});
jest.mock('../../../../src/ts/app/api/workflow/workflow-api-slice', () => {
  const actual = jest.requireActual('../../../../src/ts/app/api/workflow/workflow-api-slice');

  const mockActivityMutation = jest.fn();
  const mockAnalystStageMutation = jest.fn();

  return {
    ...actual,
    useUpdateActivityIntervalStatusMutation: () => [mockActivityMutation],
    useUpdateStageIntervalStatusMutation: () => [mockAnalystStageMutation],
    useWorkflowQuery: jest.fn(() => ({
      isSuccess: true,
      data: {
        stages: [
          {
            name: openIntervalName,
            mode: WorkflowTypes.StageMode.INTERACTIVE,
            activities: [
              {
                name: activityInterval.intervalId.definitionId.name,
                stationGroup: { name: 'mockStationGroup' }
              }
            ]
          },
          {
            name: 'AL2',
            mode: WorkflowTypes.StageMode.AUTOMATIC,
            activities: [
              {
                name: activityInterval.intervalId.definitionId.name,
                stationGroup: { name: 'mockStationGroup2' }
              }
            ]
          }
        ]
      }
    }))
  };
});
const operationalTimeRange: CommonTypes.TimeRange = {
  startTimeSecs: 0,
  endTimeSecs: 2000
};
const operationalTimePeriodConfigurationQuery = cloneDeep(useQueryStateResult);
operationalTimePeriodConfigurationQuery.data = operationalTimeRange;
jest.mock(
  '../../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice',
  () => {
    const actual = jest.requireActual(
      '../../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice'
    );

    return {
      // ...actual,
      useGetOperationalTimePeriodConfigurationQuery: jest.fn(
        () => operationalTimePeriodConfigurationQuery
      ),
      useGetProcessingAnalystConfigurationQuery: jest.fn(() => ({
        data: processingAnalystConfigurationData
      })),
      processingConfigurationApiSlice: {
        ...actual.processingConfigurationApiSlice,
        endpoints: {
          getProcessingAnalystConfiguration: {
            select: jest.fn(() =>
              jest.fn(() => ({
                data: processingAnalystConfigurationData
              }))
            )
          }
        }
      }
    };
  }
);
jest.mock('../../../../src/ts/app/hooks/workflow-hooks', () => {
  return {
    useStageId: jest.fn().mockReturnValue({
      startTime: 0,
      definitionId: {
        name: 'AL1'
      }
    })
  };
});
jest.mock('../../../../src/ts/app/hooks/station-definition-hooks', () => {
  const actual = jest.requireActual('../../../../src/ts/app/hooks/station-definition-hooks');
  return {
    ...actual,
    useVisibleStations: jest.fn(() => defaultStations)
  };
});

jest.mock('../../../../src/ts/app/hooks/station-definition-hooks', () => {
  const actual = jest.requireActual('../../../../src/ts/app/hooks/station-definition-hooks');
  return {
    ...actual,
    useVisibleStations: jest.fn(() => defaultStations)
  };
});

const mockStoreCreator: MockStoreCreator<AppState, AnyAction> = createMockStore([thunk]);

const mockAppState: AppState = {
  ...appState,
  app: {
    ...appState.app,
    analyst: {
      ...appState.app.analyst,
      currentPhase: 'S',
      openEventId: eventData2.id,
      phaseSelectorFavorites: {
        'Seismic & Hydroacoustic': ['S']
      }
    },
    workflow: { ...appState.app.workflow, openIntervalName: 'AL1' }
  },
  data: {
    ...appState.data,
    events: { [eventData2.id]: eventData2 },
    rotationTemplates: rotationTemplateByPhaseByStationRecord,
    queries: {
      ...appState.data.queries,
      getRotationTemplates: {
        AKASG: {
          '0': {
            arg: { stations: [akasgVersionReference], phases: ['S'] },
            error: undefined,
            status: AsyncActionStatus.fulfilled,
            time: 0,
            attempts: 1
          }
        }
      }
    }
  }
};
const store = mockStoreCreator(mockAppState);

const createMockOpenInterval = () => {
  store.dispatch(workflowActions.setOpenActivityNames(['AL1 Event Review']));
  store.dispatch(workflowActions.setOpenIntervalName('AL1'));
  store.dispatch(
    userSessionActions.setAuthenticationStatus({
      userName: 'TestUser',
      authenticated: true,
      authenticationCheckComplete: true,
      failedToConnect: false
    })
  );
  store.dispatch(
    workflowActions.setTimeRange({ startTimeSecs: 1669150800, endTimeSecs: 1669151800 })
  );
};

describe('useRotate2dForStations', () => {
  createMockOpenInterval();

  it('is defined', () => {
    expect(useRotate2dForStations).toBeDefined();
  });
  it('can be called given stations and a phase', async () => {
    jest.clearAllMocks();
    const result = renderHook(useRotate2dForStations, {
      wrapper: getTestReduxWrapper(store)
    });
    const akasg = defaultStations[1];

    await result.result.current([akasg], 'S');
    expect(useGetOrFetchRotationTemplateForStationAndPhase).toHaveBeenCalled();
    const mockGetOrFetchRotationTemplateForStationAndPhase: jest.Mock =
      useGetOrFetchRotationTemplateForStationAndPhase() as any;
    expect(mockGetOrFetchRotationTemplateForStationAndPhase.mock.calls[0][0]).toMatchSnapshot();
  });
  describe('validation checks for rotate2dForChannels', () => {
    createMockOpenInterval();
    it('called with provided leadDuration and Duration', async () => {
      jest.clearAllMocks();
      const result = renderHook(useRotate2dForStations, {
        wrapper: getTestReduxWrapper(store)
      });
      const akasg = defaultStations[1];
      const providedStations = [akasg];
      const providedPhase = `S`;
      const providedLeadDuration = 789234;
      const providedDuration = 908345890;

      await result.result.current(
        providedStations,
        providedPhase,
        undefined,
        providedLeadDuration,
        providedDuration
      );
      expect(useGetOrFetchRotationTemplateForStationAndPhase).toHaveBeenCalled();
      const mockGetOrFetchRotationTemplateForStationAndPhase: jest.Mock =
        useGetOrFetchRotationTemplateForStationAndPhase() as any;
      expect(mockGetOrFetchRotationTemplateForStationAndPhase).toHaveBeenCalled();
    });
    it('called with duration and leadDuration from the template when none have been provided', async () => {
      jest.clearAllMocks();
      const result = renderHook(useRotate2dForStations, {
        wrapper: getTestReduxWrapper(store)
      });
      const akasg = defaultStations[1];
      const providedStations = [akasg];
      const providedPhase = `S`;

      await result.result.current(providedStations, providedPhase);
      expect(useGetOrFetchRotationTemplateForStationAndPhase).toHaveBeenCalled();
      const mockGetOrFetchRotationTemplateForStationAndPhase: jest.Mock =
        useGetOrFetchRotationTemplateForStationAndPhase() as any;
      expect(mockGetOrFetchRotationTemplateForStationAndPhase).toHaveBeenCalled();
    });
  });
});
