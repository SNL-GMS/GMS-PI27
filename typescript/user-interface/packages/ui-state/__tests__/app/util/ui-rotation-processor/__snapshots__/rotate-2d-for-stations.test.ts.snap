// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRotate2dForStations can be called given stations and a phase 1`] = `
{
  "allRawChannels": [
    {
      "canonicalName": "AKASG.AKBB.BHE",
      "channelBandType": "HIGH_BROADBAND",
      "channelDataType": "SEISMIC",
      "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
      "channelOrientationCode": "E",
      "channelOrientationType": "EAST_WEST",
      "configuredInputs": [],
      "description": "Raw Channel created from ReferenceChannel 55ec5a5c-0280-3d57-9128-87dce456e38b with version 2b8715d3-f39b-3a36-adff-6ec262f565d1",
      "effectiveAt": 1636503404,
      "effectiveUntil": 1660701599.984,
      "location": {
        "depthKm": 0.035,
        "elevationKm": 0.16,
        "latitudeDegrees": 50.7012,
        "longitudeDegrees": 29.2242,
      },
      "name": "AKASG.AKBB.BHE",
      "nominalSampleRateHz": 100,
      "orientationAngles": {
        "horizontalAngleDeg": 90,
        "verticalAngleDeg": 90,
      },
      "processingDefinition": {},
      "processingMetadata": {
        "CHANNEL_GROUP": "AKBB",
      },
      "station": {
        "name": "AKASG",
      },
      "units": "NANOMETERS_PER_COUNT",
    },
    {
      "canonicalName": "AKASG.AKASG.BHN",
      "channelBandType": "HIGH_BROADBAND",
      "channelDataType": "SEISMIC",
      "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
      "channelOrientationCode": "N",
      "channelOrientationType": "NORTH_SOUTH",
      "configuredInputs": [],
      "description": "Raw Channel created from ReferenceChannel 8bd585b9-7795-3d62-965f-bb0e9b965b65 with version 39f21caf-a86a-3c24-94c0-994c9b89b98a",
      "effectiveAt": 1636503404,
      "effectiveUntil": 1660701599.984,
      "location": {
        "depthKm": 0.035,
        "elevationKm": 0.16,
        "latitudeDegrees": 50.7012,
        "longitudeDegrees": 29.2242,
      },
      "name": "AKASG.AKASG.BHN",
      "nominalSampleRateHz": 40,
      "orientationAngles": {
        "horizontalAngleDeg": 0,
        "verticalAngleDeg": 90,
      },
      "processingDefinition": {},
      "processingMetadata": {
        "CHANNEL_GROUP": "AKBB",
      },
      "station": {
        "name": "AKASG",
      },
      "units": "NANOMETERS_PER_COUNT",
    },
    {
      "canonicalName": "AKASG.AKASG.BHZ",
      "channelBandType": "HIGH_BROADBAND",
      "channelDataType": "SEISMIC",
      "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
      "channelOrientationCode": "Z",
      "channelOrientationType": "VERTICAL",
      "configuredInputs": [],
      "description": "Raw Channel created from ReferenceChannel fa098cdc-0f58-30bd-88fa-faadaf40b671 with version 4a7ac51e-d21d-3443-883a-f1f444f08d16",
      "effectiveAt": 1636503404,
      "effectiveUntil": 1660701599.984,
      "location": {
        "depthKm": 0.035,
        "elevationKm": 0.16,
        "latitudeDegrees": 50.7012,
        "longitudeDegrees": 29.2242,
      },
      "name": "AKASG.AKASG.BHZ",
      "nominalSampleRateHz": 40,
      "orientationAngles": {
        "horizontalAngleDeg": -1,
        "verticalAngleDeg": 0,
      },
      "processingDefinition": {},
      "processingMetadata": {
        "CHANNEL_GROUP": "AKBB",
      },
      "station": {
        "name": "AKASG",
      },
      "units": "NANOMETERS_PER_COUNT",
    },
  ],
  "channelGroups": [
    {
      "channels": [
        {
          "canonicalName": "AKASG.AKBB.BHE",
          "channelBandType": "HIGH_BROADBAND",
          "channelDataType": "SEISMIC",
          "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
          "channelOrientationCode": "E",
          "channelOrientationType": "EAST_WEST",
          "configuredInputs": [],
          "description": "Raw Channel created from ReferenceChannel 55ec5a5c-0280-3d57-9128-87dce456e38b with version 2b8715d3-f39b-3a36-adff-6ec262f565d1",
          "effectiveAt": 1636503404,
          "effectiveUntil": 1660701599.984,
          "location": {
            "depthKm": 0.035,
            "elevationKm": 0.16,
            "latitudeDegrees": 50.7012,
            "longitudeDegrees": 29.2242,
          },
          "name": "AKASG.AKBB.BHE",
          "nominalSampleRateHz": 100,
          "orientationAngles": {
            "horizontalAngleDeg": 90,
            "verticalAngleDeg": 90,
          },
          "processingDefinition": {},
          "processingMetadata": {
            "CHANNEL_GROUP": "AKBB",
          },
          "station": {
            "name": "AKASG",
          },
          "units": "NANOMETERS_PER_COUNT",
        },
        {
          "canonicalName": "AKASG.AKASG.BHN",
          "channelBandType": "HIGH_BROADBAND",
          "channelDataType": "SEISMIC",
          "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
          "channelOrientationCode": "N",
          "channelOrientationType": "NORTH_SOUTH",
          "configuredInputs": [],
          "description": "Raw Channel created from ReferenceChannel 8bd585b9-7795-3d62-965f-bb0e9b965b65 with version 39f21caf-a86a-3c24-94c0-994c9b89b98a",
          "effectiveAt": 1636503404,
          "effectiveUntil": 1660701599.984,
          "location": {
            "depthKm": 0.035,
            "elevationKm": 0.16,
            "latitudeDegrees": 50.7012,
            "longitudeDegrees": 29.2242,
          },
          "name": "AKASG.AKASG.BHN",
          "nominalSampleRateHz": 40,
          "orientationAngles": {
            "horizontalAngleDeg": 0,
            "verticalAngleDeg": 90,
          },
          "processingDefinition": {},
          "processingMetadata": {
            "CHANNEL_GROUP": "AKBB",
          },
          "station": {
            "name": "AKASG",
          },
          "units": "NANOMETERS_PER_COUNT",
        },
        {
          "canonicalName": "AKASG.AKASG.BHZ",
          "channelBandType": "HIGH_BROADBAND",
          "channelDataType": "SEISMIC",
          "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
          "channelOrientationCode": "Z",
          "channelOrientationType": "VERTICAL",
          "configuredInputs": [],
          "description": "Raw Channel created from ReferenceChannel fa098cdc-0f58-30bd-88fa-faadaf40b671 with version 4a7ac51e-d21d-3443-883a-f1f444f08d16",
          "effectiveAt": 1636503404,
          "effectiveUntil": 1660701599.984,
          "location": {
            "depthKm": 0.035,
            "elevationKm": 0.16,
            "latitudeDegrees": 50.7012,
            "longitudeDegrees": 29.2242,
          },
          "name": "AKASG.AKASG.BHZ",
          "nominalSampleRateHz": 40,
          "orientationAngles": {
            "horizontalAngleDeg": -1,
            "verticalAngleDeg": 0,
          },
          "processingDefinition": {},
          "processingMetadata": {
            "CHANNEL_GROUP": "AKBB",
          },
          "station": {
            "name": "AKASG",
          },
          "units": "NANOMETERS_PER_COUNT",
        },
      ],
      "description": "MALIN_ARRAY,_Ukraine",
      "effectiveAt": 1636503404,
      "effectiveUntil": 1660701599.984,
      "location": {
        "depthKm": 0.037,
        "elevationKm": 0.16,
        "latitudeDegrees": 50.6911,
        "longitudeDegrees": 29.2131,
      },
      "name": "AKASG",
      "type": "SITE_GROUP",
    },
  ],
  "description": "AKASG,_Tajikistan",
  "effectiveAt": 1636503404,
  "effectiveUntil": 1660701599.984,
  "location": {
    "depthKm": 0,
    "elevationKm": 2.312,
    "latitudeDegrees": 37.53,
    "longitudeDegrees": 71.66,
  },
  "name": "AKASG",
  "relativePositionsByChannel": {
    "AKASG.AKASG.HHE": {
      "eastDisplacementKm": 0,
      "northDisplacementKm": 0,
      "verticalDisplacementKm": 0,
    },
    "AKASG.AKASG.HHN": {
      "eastDisplacementKm": 0,
      "northDisplacementKm": 0,
      "verticalDisplacementKm": 0,
    },
    "AKASG.AKASG.HHZ": {
      "eastDisplacementKm": 0,
      "northDisplacementKm": 0,
      "verticalDisplacementKm": 0,
    },
  },
  "type": "SEISMIC_3_COMPONENT",
}
`;
